### PR TITLE
use rust_assert in dead code pass instead of rust_error_at

### LINKED ISSUE #550: DeadCode pass error handling
The dead code pass is used to generate warnings for relevant code. It should not produce errors I think we could take a look at using rust_assert which should be more useful in diagnosing issues with the pass. Take for example:

https://github.com/Rust-GCC/gccrs/blob/210ae4f7b0fea9671482b8f01354fd5b9274f878/gcc/rust/lint/rust-lint-marklive.cc#L237

This function wraps lookup_node_to_hir which creates an interface to ignore errors that might provide bad errors such as:

https://github.com/Rust-GCC/gccrs/blob/210ae4f7b0fea9671482b8f01354fd5b9274f878/gcc/rust/lint/rust-lint-marklive.cc#L116

if node_id_to_hir_id fails the code will continue on to mark the node anyway regardless.

let's remove this function and call lookup_node_to_hir directly where required and we can use the error handling already provided and then use a rust_assert might be the simplest solution, since there should not be any issues with resolving nodes at this stage of the pipeline. If an error did occur it would signify something else going wrong.

### PR DESCRIPTION
**Pull request by thomasyonug**
use rust_assert in dead code pass instead of rust_error_at

Fixed #550

### CODE REVIEW

**Comment by thomasyonug** [comment]
Please let me know if I got the point entirely. :)

**Comment by dkm** [comment]
Not sure this is correct. Assertion can be disabled, so any expr used in them should not do anything (no side effect) or the build will not be working with assertion disabled.

```

/* Use gcc_assert(EXPR) to test invariants.  */
#if ENABLE_ASSERT_CHECKING
#define gcc_assert(EXPR)                                                \
   ((void)(!(EXPR) ? fancy_abort (__FILE__, __LINE__, __FUNCTION__), 0 : 0))
#elif (GCC_VERSION >= 4005)
#define gcc_assert(EXPR)                                                \
  ((void)(__builtin_expect (!(EXPR), 0) ? __builtin_unreachable (), 0 : 0))
#else
/* Include EXPR, so that unused variable warnings do not occur.  */
#define gcc_assert(EXPR) ((void)(0 && (EXPR)))
#endif
```

`rust_assert` is simply a macro for `gcc_assert`

**Comment by dkm** [comment]
Not sure this is correct, as assertion can be disabled and any side effect in the expr will be removed.
```
/* Use gcc_assert(EXPR) to test invariants.  */
#if ENABLE_ASSERT_CHECKING
#define gcc_assert(EXPR)                                                \
   ((void)(!(EXPR) ? fancy_abort (__FILE__, __LINE__, __FUNCTION__), 0 : 0))
#elif (GCC_VERSION >= 4005)
#define gcc_assert(EXPR)                                                \
  ((void)(__builtin_expect (!(EXPR), 0) ? __builtin_unreachable (), 0 : 0))
#else
/* Include EXPR, so that unused variable warnings do not occur.  */
#define gcc_assert(EXPR) ((void)(0 && (EXPR)))
#endif
```

File: gcc/rust/lint/rust-lint-marklive.cc [hunk 1]
```diff
@@ -113,12 +114,13 @@ MarkLive::visit (HIR::MethodCallExpr &expr)
 
   // node back to HIR
   HirId ref;
-  node_id_to_hir_id (expr.get_mappings ().get_crate_num (), ref_node_id, ref,
-		     expr.get_locus ());
+  rust_assert (
+    mappings->lookup_node_to_hir (expr.get_mappings ().get_crate_num (),
+				  ref_node_id, &ref));
```
Comment (dkm) [inline gcc/rust/lint/rust-lint-marklive.cc:119]: When assertion are disabled, `ref` will be uninit here

File: gcc/rust/lint/rust-lint-marklive.cc [hunk 2]
```diff
@@ -127,23 +129,18 @@ MarkLive::visit_path_segment (HIR::PathExprSegment seg)
   if (resolver->lookup_resolved_name (ast_node_id, &ref_node_id))
     {
       Resolver::Definition def;
-      if (!resolver->lookup_definition (ref_node_id, &def))
-	{
-	  rust_error_at (seg.get_locus (),
-			 "unknown reference for resolved name");
-	  return false;
-	}
+      rust_assert (resolver->lookup_definition (ref_node_id, &def));
```
Comment (dkm) [inline gcc/rust/lint/rust-lint-marklive.cc:132]: same here

**Review by dkm — CHANGES_REQUESTED** [review]

**Comment by dkm** [comment]
Better use something like 
```
ret = lookup(...);
rust_assert (ret);
...
```

File: gcc/rust/lint/rust-lint-marklive.cc (hunk 1 above)
Comment (thomasyonug) [reply gcc/rust/lint/rust-lint-marklive.cc:119]: Oops, I did't notice that.

**Comment by dkm** [comment]
In fact, this may be correct as we are not using anything older than 4005.

**Comment by thomasyonug** [comment]
> In fact, this may be correct as we are not using anything older than 4005.

Sorry, I don't know what's the 4005 means.

**Comment by dkm** [comment]
> In fact, this may be correct as we are not using anything older than 4005.

I /think/ `gcc_assert` still calls `EXPR`, only `gcc_checking_assert` skips it when disabled. But I could not find any such usage in the code base, so not sure this is really used...

**Comment by thomasyonug** [comment]
> > In fact, this may be correct as we are not using anything older than 4005.
> 
> I /think/ `gcc_assert` still calls `EXPR`, only `gcc_checking_assert` skips it when disabled. But I could not find any such usage in the code base, so not sure this is really used...

Ha, gotcha, thanks. I think it's better to avoid taking a risk.

**Comment by dkm** [comment]
> > In fact, this may be correct as we are not using anything older than 4005.
> 
> Sorry, I don't know what's the 4005 means. Would you mind explain more details to me? Or any reference

Sorry, it's the GCC version.  `4XXX` is for GCC 4, so not a gccrs target :)

File: gcc/rust/lint/rust-lint-marklive.cc [hunk 3]
```diff
@@ -113,12 +114,14 @@ MarkLive::visit (HIR::MethodCallExpr &expr)
 
   // node back to HIR
   HirId ref;
-  node_id_to_hir_id (expr.get_mappings ().get_crate_num (), ref_node_id, ref,
-		     expr.get_locus ());
+  bool ret
```
Comment (philberty) [inline gcc/rust/lint/rust-lint-marklive.cc:117]: usually we do bool ok = mappings->lookup_node_to_hir

I think 'ok' reads better

File: gcc/rust/lint/rust-lint-marklive.h [hunk 4]
```diff
@@ -266,10 +266,9 @@ class MarkLive : public MarkLiveBase
       tyctx (Resolver::TypeCheckContext::get ()){};
 
   void mark_hir_id (HirId);
-  bool visit_path_segment (HIR::PathExprSegment);
```
Comment (philberty) [inline gcc/rust/lint/rust-lint-marklive.h:269]: I think keep this bool visit_path_segment for now

**Review by philberty — CHANGES_REQUESTED** [review]

File: gcc/rust/lint/rust-lint-marklive.h (hunk 4 above)
Comment (thomasyonug) [reply gcc/rust/lint/rust-lint-marklive.h:269]: May I ask why shoud I keep this bool for now?

File: gcc/rust/lint/rust-lint-marklive.cc (hunk 3 above)
Comment (thomasyonug) [reply gcc/rust/lint/rust-lint-marklive.cc:117]: Yeah, sure it is

**Comment by thomasyonug** [comment]
I'll squash the commits after approve

**Comment by dkm** [comment]
I guess @philberty was asking for changing all `ret` to `ok`, not only the one for node_to_hir?

**Comment by thomasyonug** [comment]
> I guess @philberty was asking for changing all `ret` to `ok`, not only the one for node_to_hir?

Yeah, I miss some of them. Thanks for pointing.

**Review by dkm — APPROVED** [review]
Looks fine :)

File: gcc/rust/lint/rust-lint-marklive.h (hunk 4 above)
Comment (philberty) [reply gcc/rust/lint/rust-lint-marklive.h:269]: it may present other options for more error handling later on.

**Review by philberty — APPROVED** [review]

**Pull request merged**